File: stateful/src/main/java/com/github/zchu/stateful/StatefulView.java
package com.github.zchu.stateful;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;
import androidx.annotation.*;


/**
 * Created by z-chu on 2016/6/8.
 */
public class StatefulView extends FrameLayout implements View.OnClickListener {

    public static final int STATE_NONE = 0;
    public static final int STATE_LOADING = 1;
    public static final int STATE_ERROR = 2;
    public static final int STATE_FINISH = 3;
    //单位dp
    private static final int ANIM_TRANSLATE_Y = 40;
    //动画持续时间
    private static final int ANIM_TIME_LONG = 500;

    private LayoutInflater mLayoutInflater;
    private View mLoadingView;
    private View mErrorView;
    private View mContentView;
    private int mContentLayoutId;
    private int mLoadingLayoutId;
    private int mErrorLayoutId;
    private int mState = STATE_NONE;

    private TextView tvLoadingMessage;
    private TextView tvErrorMessage;


    private OnRetryListener mOnRetryListener;
    private OnViewCreatedListener mContentViewCreatedListener;
    private OnViewCreatedListener mLoadingViewCreatedListener;
    private OnViewCreatedListener mErrorViewCreatedListener = new OnViewCreatedListener() {
        @Override
        public void onViewCreated(@NonNull View view) {
            View viewById = view.findViewById(R.id.reload_view);
            if (viewById != null) {
                viewById.setOnClickListener(StatefulView.this);
            }
        }
    };


    public StatefulView(Context context) {
        this(context, null);
    }

    public StatefulView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public StatefulView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        inflateView(context, context.obtainStyledAttributes(attrs, R.styleable.StatefulView, defStyleAttr, 0));
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public StatefulView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        inflateView(context, context.obtainStyledAttributes(attrs, R.styleable.StatefulView, defStyleAttr, defStyleRes));

    }

    private void inflateView(Context context, TypedArray typedArray) {
        mLayoutInflater = LayoutInflater.from(context);
        mContentLayoutId = typedArray.getResourceId(R.styleable.StatefulView_contentLayout, 0);
        mLoadingLayoutId = typedArray.getResourceId(R.styleable.StatefulView_loadingLayout, 0);
        mErrorLayoutId = typedArray.getResourceId(R.styleable.StatefulView_errorLayout, 0);
        typedArray.recycle();
    }


    public void showLoading() {
        showLoading(null);
    }

    public void showLoading(@Nullable String message) {
        if (mState == STATE_LOADING) {
            return;
        }
        this.mState = STATE_LOADING;
        if (mLoadingView != null) {
            mLoadingView.setVisibility(VISIBLE);
        } else {
            if (mLoadingLayoutId != 0) {
                mLoadingView = mLayoutInflater.inflate(mLoadingLayoutId, this, false);
                addView(mLoadingView);
                if (mLoadingViewCreatedListener != null) {
                    mLoadingViewCreatedListener.onViewCreated(mLoadingView);
                }
            }
        }
        if (message != null && mLoadingView != null) {
            if (tvLoadingMessage == null) {
                tvLoadingMessage = mLoadingView.findViewById(R.id.tv_loading_message);

            }
            if (tvLoadingMessage != null) {
                tvLoadingMessage.setVisibility(View.VISIBLE);
                tvLoadingMessage.setText(message);
            }
        }
        if (mErrorView != null) {
            mErrorView.setVisibility(GONE);
        }

    }

    public void showLoading(@StringRes int resId) {
        showLoading(getContext().getString(resId));
    }

    public void showError() {
        showError(null);
    }

    public void showError(@StringRes int resId) {
        showError(getContext().getString(resId));
    }


    public void showError(@Nullable String message) {
        if (mState == STATE_ERROR) {
            return;
        }
        this.mState = STATE_ERROR;
        if (mLoadingView != null) {
            mLoadingView.setVisibility(GONE);
        }
        if (mErrorView != null) {
            mErrorView.setVisibility(VISIBLE);
        } else {
            if (mErrorLayoutId != 0) {
                mErrorView = mLayoutInflater.inflate(mErrorLayoutId, this, false);
                addView(mErrorView);
                if (mErrorViewCreatedListener != null) {
                    mErrorViewCreatedListener.onViewCreated(mErrorView);
                }
                // mErrorView.findViewById(R.id.reload_view).setOnClickListener(this);
            }
        }
        if (message != null && mErrorView != null) {
            if (tvErrorMessage == null) {
                tvErrorMessage = mErrorView.findViewById(R.id.tv_error_message);
            }
            if (tvErrorMessage != null) {
                tvErrorMessage.setVisibility(View.VISIBLE);
                tvErrorMessage.setText(message);
            }
        }
    }


    public void showContent() {
        showContent(true);
    }

    public void showContent(boolean anim) {
        if (mState == STATE_FINISH) {
            return;
        }
        this.mState = STATE_FINISH;
        if (mErrorView != null) {
            mErrorView.setVisibility(View.GONE);
        }

        if (mContentView != null || mContentLayoutId != 0) {
            if (mContentView == null) {
                mContentView = mLayoutInflater.inflate(mContentLayoutId, this, false);
                addView(mContentView, 0);
                if (mContentViewCreatedListener != null) {
                    mContentViewCreatedListener.onViewCreated(mContentView);
                }
            }
            mContentView.setVisibility(View.VISIBLE);
            if (anim) {
                startShowContentAnim(mContentView, mLoadingView);
            } else {
                if (mLoadingView != null) {
                    mLoadingView.setVisibility(View.GONE);
                }
            }
        } else {
            if (mLoadingView != null) {
                mLoadingView.setVisibility(View.GONE);
            }
        }

    }

    private void startShowContentAnim(View contentView, final View loadingView) {
        AnimatorSet animatorSet = new AnimatorSet();
        if (contentView != null && contentView.getVisibility() == View.VISIBLE) {
            ObjectAnimator contentFadeIn = ObjectAnimator.ofFloat(contentView, View.ALPHA, 0f, 1f);
            ObjectAnimator contentTranslateIn = ObjectAnimator.ofFloat(contentView, View.TRANSLATION_Y,
                    ANIM_TRANSLATE_Y, 0);
            animatorSet.playTogether(contentFadeIn, contentTranslateIn);
        }

        if (loadingView != null && loadingView.getVisibility() == View.VISIBLE) {
            ObjectAnimator loadingFadeOut = ObjectAnimator.ofFloat(loadingView, View.ALPHA, 1f, 0f);
            ObjectAnimator loadingTranslateOut = ObjectAnimator.ofFloat(loadingView, View.TRANSLATION_Y, 0,
                    -ANIM_TRANSLATE_Y * 2);
            animatorSet.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    loadingView.setVisibility(View.GONE);
                    loadingView.setAlpha(1f); // For future showLoading calls
                    loadingView.setTranslationY(0);
                }

                @Override
                public void onAnimationStart(Animator animation) {
                }
            });
            animatorSet.playTogether(loadingFadeOut, loadingTranslateOut);

        }
        animatorSet.setDuration(ANIM_TIME_LONG);
        animatorSet.start();
    }

    public void onClick(View v) {
        if (mOnRetryListener != null) {
            //点击重新加载
            mOnRetryListener.onRetry(v);
        }
    }

    public void setLoadingLayoutId(@LayoutRes int layoutId) {
        checkIsLegalStatus();
        this.mLoadingLayoutId = layoutId;
    }

    public void setLoadingView(View view) {
        checkIsLegalStatus();
        if (view != null) {
            this.mLoadingView = view;
            addView(mLoadingView);
            if (mLoadingViewCreatedListener != null) {
                mLoadingViewCreatedListener.onViewCreated(mLoadingView);
            }

            mLoadingView.setVisibility(View.GONE);
        }
    }

    public void setErrorLayoutId(int layoutId) {
        checkIsLegalStatus();
        this.mErrorLayoutId = layoutId;
    }

    public void setErrorView(View view) {
        checkIsLegalStatus();
        if (view != null) {
            this.mErrorView = view;
            addView(mErrorView);
            if (mErrorViewCreatedListener != null) {
                mErrorViewCreatedListener.onViewCreated(mErrorView);
            }

            mErrorView.setVisibility(View.GONE);
        }
    }

    public void setContentLayoutId(int layoutId) {
        checkIsLegalStatus();
        this.mContentLayoutId = layoutId;
    }

    public void setContentView(View view) {
        checkIsLegalStatus();
        if (view != null) {
            this.mContentView = view;
            addView(mContentView, 0);
            if (mContentViewCreatedListener != null) {
                mContentViewCreatedListener.onViewCreated(mContentView);
            }
            mContentView.setVisibility(View.GONE);
        }
    }

    private void checkIsLegalStatus() {
        if (mState != STATE_NONE) {
            throw new IllegalStateException("Can not change view , because" + getClass().getSimpleName() + "\'s state is not STATE_NONE");
        }
    }


    public int getState() {
        return mState;
    }

    /**
     * 设置点击重试的回调，注意如果还设置ErrorViewCreatedListener，那setOnRetryListener设置的listener将会失效
     */
    public void setOnRetryListener(OnRetryListener listener) {
        this.mOnRetryListener = listener;
    }

    public void setLoadingViewCreatedListener(OnViewCreatedListener listener) {
        if (mLoadingView != null && mLoadingView.getParent() == this) {
            listener.onViewCreated(mContentView);
        } else {
            this.mLoadingViewCreatedListener = listener;
        }
    }

    public void setContentViewCreatedListener(OnViewCreatedListener listener) {
        if (mContentView != null && mContentView.getParent() == this) {
            listener.onViewCreated(mContentView);
        } else {
            this.mContentViewCreatedListener = listener;
        }
    }

    /**
     * 设置ErrorView 在创建时回调，注意设置完之后，setOnRetryListener设置的listener将会失效
     */
    public void setErrorViewCreatedListener(OnViewCreatedListener listener) {
        if (mErrorView != null && mErrorView.getParent() == this) {
            listener.onViewCreated(mErrorView);
        } else {
            this.mErrorViewCreatedListener = listener;
        }
    }

    public interface OnRetryListener {
        /**
         * 重新加载
         */
        void onRetry(View view);
    }

    public interface OnViewCreatedListener {
        void onViewCreated(@NonNull View view);
    }

}

File: statefulView/src/main/java/com/github/zchu/stateful/StatefulView.kt
package com.github.zchu.stateful

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.content.res.TypedArray
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.annotation.*


class StatefulView : FrameLayout {

    private lateinit var mLayoutInflater: LayoutInflater
    private var mLoadingView: View? = null
        set(value) {
            if (mLoadingTextViewId != View.NO_ID) {
                tvLoadingMessage = value?.findViewById(mLoadingTextViewId)
            }
            field = value
        }

    private var mErrorView: View? = null
        set(value) {
            if (mRetryViewId != View.NO_ID) {
                val retryView = value?.findViewById<View>(mRetryViewId)
                retryView?.setOnClickListener {
                    onRetryListener?.invoke(it)
                }
            }
            if (mErrorTextViewId != View.NO_ID) {
                tvErrorMessage = value?.findViewById(mErrorTextViewId)
            }
            field = value
        }
    private var mContentView: View? = null

    private var mContentLayoutId: Int = View.NO_ID
        set(value) {
            if (lazyLoading && value != View.NO_ID) {
                val contentView = mLayoutInflater.inflate(mContentLayoutId, this, false)
                mContentView = contentView
                addView(contentView, 0)
                onContentViewCreatedListener?.invoke(contentView)
            }
            field = value
        }

    private var mLoadingLayoutId: Int = View.NO_ID
        set(value) {
            if (lazyLoading && value != View.NO_ID) {
                val loadingView = mLayoutInflater.inflate(mLoadingLayoutId, this, false)
                mLoadingView = loadingView
                addView(loadingView)
                onLoadingViewCreatedListener?.invoke(loadingView)
            }
            field = value
        }

    private var mErrorLayoutId: Int = View.NO_ID
        set(value) {
            if (lazyLoading && value != View.NO_ID) {
                val errorView = mLayoutInflater.inflate(mErrorLayoutId, this, false)
                mErrorView = errorView
                addView(errorView)
                onErrorViewCreatedListener?.invoke(errorView)
            }
            field = value
        }
    var state = STATE_NONE
        private set

    private var tvLoadingMessage: TextView? = null
    private var tvErrorMessage: TextView? = null
    private var mLoadingTextViewId: Int = View.NO_ID
    private var mErrorTextViewId: Int = View.NO_ID
    private var mRetryViewId: Int = View.NO_ID
    private var lazyLoading: Boolean = false


    var onRetryListener: ((View) -> Unit)? = null
    var onContentViewCreatedListener: ((View) -> Unit)? = null
        set(value) {
            mContentView?.let {
                value?.invoke(it)
            }
            field = value
        }
    var onLoadingViewCreatedListener: ((View) -> Unit)? = null
        set(value) {
            mLoadingView?.let {
                value?.invoke(it)
            }
            field = value
        }
    var onErrorViewCreatedListener: ((View) -> Unit)? = null
        set(value) {
            mErrorView?.let {
                value?.invoke(it)
            }
            field = value
        }

    @JvmOverloads
    constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        inflateView(context, context.obtainStyledAttributes(attrs, R.styleable.StatefulView, defStyleAttr, 0))
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int, @StyleRes defStyleRes: Int) : super(
        context,
        attrs,
        defStyleAttr,
        defStyleRes
    ) {
        inflateView(context, context.obtainStyledAttributes(attrs, R.styleable.StatefulView, defStyleAttr, defStyleRes))

    }

    private fun inflateView(context: Context, typedArray: TypedArray) {
        mLayoutInflater = LayoutInflater.from(context)
        lazyLoading = typedArray.getBoolean(R.styleable.StatefulView_LazyLoading, false)
        mContentLayoutId = typedArray.getResourceId(R.styleable.StatefulView_contentLayout, mContentLayoutId)
        mLoadingLayoutId = typedArray.getResourceId(R.styleable.StatefulView_loadingLayout, mLoadingLayoutId)
        mErrorLayoutId = typedArray.getResourceId(R.styleable.StatefulView_errorLayout, mErrorLayoutId)
        mErrorTextViewId = typedArray.getResourceId(R.styleable.StatefulView_errorTextViewId, mErrorTextViewId)
        mLoadingTextViewId = typedArray.getResourceId(R.styleable.StatefulView_loadingTextViewId, mLoadingTextViewId)
        mRetryViewId = typedArray.getResourceId(R.styleable.StatefulView_retryViewId, mRetryViewId)
        typedArray.recycle()

    }

    @JvmOverloads
    fun showLoading(message: String? = null) {
        if (state == STATE_LOADING) {
            return
        }
        this.state = STATE_LOADING
        var loadingView = mLoadingView
        if (loadingView != null) {
            loadingView.visibility = View.VISIBLE
        } else {
            if (mLoadingLayoutId != View.NO_ID) {
                loadingView = mLayoutInflater.inflate(mLoadingLayoutId, this, false)
                mLoadingView = loadingView
                addView(loadingView)
                onLoadingViewCreatedListener?.invoke(loadingView)
            }
        }
        if (message != null) {
            tvLoadingMessage?.let {
                it.visibility = View.VISIBLE
                it.text = message
            }
        }
        mErrorView?.visibility = View.GONE

    }

    fun showLoading(@StringRes resId: Int) {
        showLoading(context.getString(resId))
    }

    fun showError(@StringRes resId: Int) {
        showError(context.getString(resId))
    }


    @JvmOverloads
    fun showError(message: String? = null) {
        if (state == STATE_ERROR) {
            return
        }
        this.state = STATE_ERROR
        mLoadingView?.visibility = View.GONE
        var errorView = mErrorView
        if (errorView != null) {
            errorView.visibility = View.VISIBLE

        } else {
            if (mErrorLayoutId != View.NO_ID) {
                errorView = mLayoutInflater.inflate(mErrorLayoutId, this, false)
                mErrorView = errorView
                addView(errorView)
                onErrorViewCreatedListener?.invoke(errorView)
            }
        }
        if (message != null) {
            tvErrorMessage?.let {
                it.visibility = View.VISIBLE
                it.text = message
            }
        }
    }

    @JvmOverloads
    fun showContent(anim: Boolean = true) {
        if (state == STATE_FINISH) {
            return
        }
        this.state = STATE_FINISH
        mErrorView?.visibility = View.GONE
        var contentView = mContentView
        if (contentView != null) {
            contentView.visibility = View.VISIBLE
        } else {
            if (mContentLayoutId != View.NO_ID) {
                contentView = mLayoutInflater.inflate(mContentLayoutId, this, false)
                mContentView = contentView
                addView(contentView, 0)
                onContentViewCreatedListener?.invoke(contentView)
            }
        }
        contentView?.visibility = View.VISIBLE
        if (anim) {
            startShowContentAnim(contentView, mLoadingView)
        } else {
            mLoadingView?.visibility = View.GONE
        }
    }

    private fun startShowContentAnim(contentView: View?, loadingView: View?) {
        val animatorSet = AnimatorSet()
        if (contentView != null && contentView.visibility == View.VISIBLE) {
            val contentFadeIn = ObjectAnimator.ofFloat(contentView, View.ALPHA, 0f, 1f)
            val contentTranslateIn = ObjectAnimator.ofFloat<View>(
                contentView, View.TRANSLATION_Y,
                ANIM_TRANSLATE_Y, 0F
            )
            animatorSet.playTogether(contentFadeIn, contentTranslateIn)
        }

        if (loadingView != null && loadingView.visibility == View.VISIBLE) {
            val loadingFadeOut = ObjectAnimator.ofFloat(loadingView, View.ALPHA, 1f, 0f)
            val loadingTranslateOut = ObjectAnimator.ofFloat<View>(
                loadingView, View.TRANSLATION_Y, 0F,
                -ANIM_TRANSLATE_Y * 2
            )
            animatorSet.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    loadingView.visibility = View.GONE
                    loadingView.alpha = 1f // For future showLoading calls
                    loadingView.translationY = 0f
                }

                override fun onAnimationStart(animation: Animator) {}
            })
            animatorSet.playTogether(loadingFadeOut, loadingTranslateOut)

        }
        animatorSet.duration = ANIM_TIME_LONG.toLong()
        animatorSet.start()
    }

    fun setLoadingLayoutId(@LayoutRes layoutId: Int) {
        checkIsLegalStatus()
        this.mLoadingLayoutId = layoutId
        mLoadingView = null
    }

    fun setLoadingView(view: View) {
        checkIsLegalStatus()
        view.visibility = View.GONE
        this.mLoadingView = view
        addView(view)
        onLoadingViewCreatedListener?.invoke(view)
    }

    fun setErrorLayoutId(layoutId: Int) {
        checkIsLegalStatus()
        this.mErrorLayoutId = layoutId
        mErrorView = null
    }

    fun setErrorView(view: View) {
        checkIsLegalStatus()
        view.visibility = View.GONE
        this.mErrorView = view
        addView(view)
        onErrorViewCreatedListener?.invoke(view)
    }

    fun setContentLayoutId(layoutId: Int) {
        checkIsLegalStatus()
        this.mContentLayoutId = layoutId
        mContentView = null
    }

    fun setContentView(view: View) {
        checkIsLegalStatus()
        view.visibility = View.GONE
        this.mContentView = view
        addView(view, 0)
        onContentViewCreatedListener?.invoke(view)
    }

    private fun checkIsLegalStatus() {
        if (state != STATE_NONE) {
            throw IllegalStateException("Can not change view , because" + javaClass.simpleName + "\'s state is not STATE_NONE")
        }
    }


    companion object {

        val STATE_NONE = 0
        val STATE_LOADING = 1
        val STATE_ERROR = 2
        val STATE_FINISH = 3
        //单位dp
        private val ANIM_TRANSLATE_Y = 40F
        //动画持续时间
        private val ANIM_TIME_LONG = 500
    }

}
